packages/gatsby-source-filesystem/src/gatsby-node.js

```diff
@@ -1,61 +1,124 @@
 const chokidar = require(`chokidar`)
 const fs = require(`fs`)
 const path = require(`path`)
-const { Machine } = require(`xstate`)
+const { Machine, interpret } = require(`xstate`)
 
 const { createFileNode } = require(`./create-file-node`)
 
 /**
  * Create a state machine to manage Chokidar's not-ready/ready states.
  */
-const createFSMachine = () =>
-  Machine({
-    key: `emitFSEvents`,
-    parallel: true,
-    strict: true,
-    states: {
-      CHOKIDAR: {
-        initial: `CHOKIDAR_NOT_READY`,
-        states: {
-          CHOKIDAR_NOT_READY: {
-            on: {
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
-              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+const createFSMachine = (
+  { actions: { createNode, deleteNode }, getNode, createNodeId, reporter },
+  pluginOptions
+) => {
+  // For every path that is reported before the 'ready' event, we throw them
+  // into a queue and then flush the queue when 'ready' event arrives.
+  // After 'ready', we handle the 'add' event without putting it into a queue.
+  let pathQueue = []
+  const flushPathQueue = () => {
+    let queue = pathQueue.slice()
+    pathQueue = null
+    return Promise.all(queue.map(createAndProcessNode))
+  }
+
+  const createAndProcessNode = path => {
+    const fileNodePromise = createFileNode(
+      path,
+      createNodeId,
+      pluginOptions
+    ).then(fileNode => {
+      createNode(fileNode)
+      return null
+    })
+    return fileNodePromise
+  }
+
+  const fsMachine = Machine(
+    {
+      id: `fs`,
+      type: `parallel`,
+      states: {
+        BOOTSTRAP: {
+          initial: `BOOTSTRAPPING`,
+          states: {
+            BOOTSTRAPPING: {
+              on: {
+                BOOTSTRAP_FINISHED: `BOOTSTRAPPED`,
+              },
+              exit: `flushPathQueue`,
             },
-          },
-          CHOKIDAR_WATCHING: {
-            on: {
-              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
+            BOOTSTRAPPED: {
+              type: `final`,
             },
           },
-          CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED: {
-            on: {
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+        },
+        CHOKIDAR: {
+          initial: `NOT_READY`,
+          states: {
+            NOT_READY: {
+              on: {
+                CHOKIDAR_READY: `READY`,
+                CHOKIDAR_ADD: { actions: `queueNodeProcessing` },
+              },
             },
+            READY: {
+              on: {
+                CHOKIDAR_ADD: { actions: `createAndProcessNode` },
+              },
+            },
+          },
+          // TODO: those two were not restricted to READY state, but maybe they should? or even maybe it should queue in NOT_READY?
+          on: {
+            CHOKIDAR_CHANGE: { actions: `createAndProcessNode` },
```

Comment:
This can be handled in exactly same way as `CHOKIDAR_ADD` event (so queue in NOT_READY and process immediately in READY)

Reply:
Maybe even `CHOKIDAR_ADD` / `CHOKIDAR_CHANGE` event can be merged because as far as plugin is concerned there's not much difference between creating and updating nodes (gatsby core takes care of this stuff).

But I don't think that verbosity of events is any problem and this might be easier to read and understand

Reply:
Ok, so you have actually answered here my previous question 😅 Still would be in favor of keeping current logic for now, I can prepare a PR with that queuing immediately after this would get merged in. 

As to verbosity of events - I would be in favor of keeping separate names.